File: 13_roman_to_int/rom_to_int.go
package main

import (
	"fmt"
)

func romanToInt(s string) int {
	romanMap := map[string]int{"I": 1, "V": 5, "X": 10, "L": 50, "C": 100, "D": 500, "M": 1000}
	byte_str := []rune(s)
	sum := 0
	start := 0
	end := len(s) - 1
	for {
		if start > end {
			break
		}

		lhs := string(byte_str[start])
		if start == end {
			sum += romanMap[lhs]
			start++
			continue
		}
		rhs := string(byte_str[start+1])

		if lhs == rhs {
			sum += 2 * romanMap[rhs]
			start += 2
		} else {
			lhsNum := romanMap[lhs]
			rhsNum := romanMap[rhs]

			if lhsNum < rhsNum {
				sum += (rhsNum - lhsNum)
				start += 2
			} else {
				sum += lhsNum
				start++
			}
		}
	}

	return sum
}

func main() {
	fmt.Println(romanToInt("III"))
	fmt.Println(romanToInt("LVIII"))
	fmt.Println(romanToInt("MCMXCIV"))
	fmt.Println(romanToInt("IV"))
}

File: 13_roman_to_int/rom_to_int.py
class Solution:
    def romanToInt(self, s: str) -> int:
        romanDict = {"I":1, "V":5, "X":10, "L":50, "C":100, "D":500, "M":1000}

        sum = 0
        finished = False
        lhs = 0
        rhs = 1

        while not finished:
            if lhs == len(s)-1:
                sum += romanDict[s[lhs]]
                finished = True
            elif romanDict[s[lhs]] < romanDict[s[rhs]]:
                sum += romanDict[s[rhs]] - romanDict[s[lhs]]
                lhs+=2
                rhs+=2
            else:
                sum += romanDict[s[lhs]]
                lhs +=1
                rhs +=1

            if lhs >= len(s):
                finished = True

        return sum

if __name__ == "__main__":
    sol = Solution()

    s ="III"
    print(sol.romanToInt(s))

    s ="LVIII"
    print(sol.romanToInt(s))

    s ="MCMXCIV"
    print(sol.romanToInt(s))